### PR TITLE
feat: use DI for structured Jacobians

### PR DESCRIPTION
**Pull request by avik-pal**
Seems to work beautifully. Haven't benchmarked yet, but the important part here is to ensure structure is preserved, since the main cost in these workloads is that Linear solving will be bottlnecked if a banded matrix is converted into a sparse matrix.

### CODE REVIEW

**Comment by gdalle** [comment]
Structure is indeed preserved for decompression (on the SparseMatrixColorings [PR branch](https://github.com/gdalle/SparseMatrixColorings.jl/pull/132) at least).
However, with the current implementation, a sparse matrix and its transpose _will be allocated_ from the structured Jacobian nonetheless, as a preparation step for decompression.
This seems futile when the decompression target still has the same structure, but when it doesn't, this sparse encoding is actually useful for fast decompression. The scenario here would be someone passing `sparsity_pattern::Tridiagonal` but then calling `jacobian` on a `jac_prototype::SparseMatrixCSC`.
Ultimately, coloring is a one-time cost, whereas decompression is paid at every Jacobian computation. So in SparseMatrixColorings we choose to allocate a little more during coloring, and that way we speed up more decompression scenarios down the road.

File: src/internal/jacobian.jl [hunk 1]
```diff
@@ -126,20 +103,19 @@ function JacobianCache(prob, alg, f::F, fu_, u, p; stats, autodiff = nothing,
...
-        J, f, fu, u, p, stats, autodiff, di_extras, sdifft_extras)
+    return JacobianCache{iip}(J, f, fu, u, p, stats, autodiff, di_extras)
 end
 
 function JacobianCache(prob, alg, f::F, ::Number, u::Number, p; stats,
         autodiff = nothing, kwargs...) where {F}
     fu = f(u, p)
     if SciMLBase.has_jac(f) || SciMLBase.has_vjp(f) || SciMLBase.has_jvp(f)
-        return JacobianCache{false}(u, f, fu, u, p, stats, autodiff, nothing, nothing)
+        return JacobianCache{false}(u, f, fu, u, p, stats, autodiff, nothing)
```
Comment (gdalle) [inline src/internal/jacobian.jl:113]: Same question about cases where DI preparation is skipped

File: src/internal/jacobian.jl [hunk 2]
```diff
@@ -63,31 +62,13 @@ function JacobianCache(prob, alg, f::F, fu_, u, p; stats, autodiff = nothing,
...
-            di_extras = if iip
-                DI.prepare_jacobian(f, fu, autodiff, u, Constant(prob.p))
-            else
-                DI.prepare_jacobian(f, autodiff, u, Constant(prob.p))
-            end
+            DI.prepare_jacobian(f, autodiff, u, Constant(prob.p))
         end
     else
-        using_sparsedifftools = false
         di_extras = nothing
```
Comment (gdalle) [inline src/internal/jacobian.jl:71]: In which cases do we not prepare the jacobian? With DI, if you forget to do that, you'll recompute the sparsity pattern, the coloring and the basis vectors every single time

Comment (avik-pal) [reply src/internal/jacobian.jl:71]: accidental, removed

File: src/internal/jacobian.jl (hunk 1 above)
Comment (avik-pal) [reply src/internal/jacobian.jl:113]: Only where the user provides an analytic jacobian. This one is a special case for scalars when DI is skipped if jacobian/jvp/vjp is provided.

File: src/internal/jacobian.jl (hunk 2 above)
Comment (oscardssmith) [reply src/internal/jacobian.jl:71]: you don't prepare the jacobian if the solver doesn't need the jacobian and the user didn't provide a jacobian function. In these cases, the nonlinear solver is doing everything with vjps and jvps anyway so you don't need this info.

**Pull request merged**